### DISCUSSION TITLE
ERROR: Cannot link_whole a custom or Rust target 'curl' into a static library 'sgx_util'

### DISCUSSION THREAD

**Discussion by LeoJungleKing** [gramineproject/gramine#1501] [Q&A] [answered]
Hello,

Ubuntu 20.04
Linux 5.15.0-1042-azure

I tried to build Gramine 1.4 with meson like that:

`meson setup build/ --buildtype=release -Ddirect=enabled -Dsgx=enabled -Ddcap=enabled -Dsgx_driver=upstream`

I succeeded in executing:

- tomlc99
- uthash-2.1.0
- mbedtls-mbedtls-3.3.0
- curl-7.84.0
- cJSON-1.7.12

But after the cJSON execution I've the following error (I let the cJSON execution to show that it passed):

```
...

Executing subproject cJSON-1.7.12 


cJSON-1.7.12| Project name: cJSON
cJSON-1.7.12| Project version: undefined
cJSON-1.7.12| C compiler for the host machine: cc (gcc 9.4.0 "cc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0")
cJSON-1.7.12| C linker for the host machine: cc ld.bfd 2.36.1
cJSON-1.7.12| Build targets in project: 7
cJSON-1.7.12| Subproject cJSON-1.7.12 finished.

Program protoc-c found: YES (/usr/bin/protoc-c)
Run-time dependency threads found: YES
Found pkg-config: /usr/bin/pkg-config (1.6.3)
Found CMake: /usr/local/bin/cmake (3.27.3)
Run-time dependency cjson found: NO (tried pkgconfig and cmake)
Library cjson found: YES
Run-time dependency libprotobuf-c found: YES 1.3.3
Library sgx_dcap_quoteverify found: YES
Configuring host_sgx_driver.h using configuration
Compiler for C supports arguments -Wno-gnu-alignof-expression: NO 
Configuring __init__.py using configuration
Configuring gramine-direct using configuration

tools/sgx/common/meson.build:1:11: ERROR: Cannot link_whole a custom or Rust target 'curl' into a 
static library 'sgx_util'. 
Instead, pass individual object files with the "objects:" keyword argument if possible. 
Meson had to promote link to link_whole because 'sgx_util' is installed but not 'curl', and thus has to 
include objects from 'curl' to be usable.

```

Do you have any ideas how to fix it?

**Comment by dimakuv** [comment, marked as answer]
Yes, this is a known problem with the latest Meson version (v1.2.0). See the corresponding explanations:
- https://github.com/gramineproject/gramine/issues/1459
- https://github.com/gramineproject/gramine/pull/1472

Unfortunately, Gramine build scripts had a bug that was exposed in meson v1.2.0.

So you have two choices:
- Either use the latest Gramine commit from the master branch (v1.4 and even v1.5 won't work)
- Or downgrade your meson to anything lower than v1.2.0 (e.g. v1.1.0).

**Reply by LeoJungleKing** [reply]
Perfect!

Thanks a lot dimakuv.
